src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.targets

```diff
@@ -30,6 +30,7 @@ The .NET Foundation licenses this file to you under the MIT license.
     <RunILLink>false</RunILLink>
     <_IsiOSLikePlatform Condition="'$(_targetOS)' == 'maccatalyst' or $(_targetOS.StartsWith('ios')) or $(_targetOS.StartsWith('tvos'))">true</_IsiOSLikePlatform>
     <_IsApplePlatform Condition="'$(_targetOS)' == 'osx' or '$(_IsiOSLikePlatform)' == 'true'">true</_IsApplePlatform>
+    <_iOSLibraryMode Condition="'$(_IsiOSLikePlatform)' == 'true' and !$(TargetFramework.Contains('$(_targetOS)')) and '$(NativeLib)' != '' and '$(CustomNativeMain)' != 'true'">true</_iOSLibraryMode>
```

Comment:
Library Mode term sounds too special in this context. Can this condition be simplified to `'$(_IsiOSLikePlatform)' == 'true' && '$(NativeLib)' == 'Shared'` and inline it at call-site?  `static` and `static-pie` libs are a bit different.

Reply:
Is there a better umbrella name to refer to when we want to consume a .net library from native?

Reply:
With `<OutputType>Exe` and `<PublishAot>true` in csproj, dotnet-publish generates a native executable. With `<OutputType>Shared` and `<PublishAot>true` we get  a shared library. We can additionally set `<NativeLib>Static` to get an static lib.

For platform specialization, we have targetXX and isXX type properties in place. Unless "Library Mod"e in `nativeaot/BuildIntegration` context is bringing something very unique (which, going by the first bullet in top comment, doesn't seem to be the case), it's better to use the existing / nearby convention.

Reply:
With `<OutputType>Exe` and `<PublishAot>true` in csproj, dotnet-publish generates a native executable. With `<OutputType>Library` and `<PublishAot>true` we get  a shared library. We can additionally set `<NativeLib>Static` to get a static lib.

For platform specialization, we have targetXX and isXX type properties in place. Unless "Library Mod"e in `nativeaot/BuildIntegration` context is bringing something very unique (which, going by the first bullet in top comment, doesn't seem to be the case), it's better to use the existing / nearby convention.

Reply:
I agree with what @jkotas and @am11 are saying here. `_iOSLibraryMode` is introducing a new concept and the question is if a new concept is avoidable.

The way we typically do checks here is that we spell out if something is platform specific or library mode specific. Like so:

https://github.com/dotnet/runtime/blob/6f4f268d0b46fa5dbcee46b7220919defb148127/src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.targets#L384

https://github.com/dotnet/runtime/blob/6f4f268d0b46fa5dbcee46b7220919defb148127/src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.targets#L69

(Plenty of other examples. We don't introduce a new property because the condition is still concise and readable.)

The conditions when `_iOSLibraryMode` is true are really difficult to parse for me. I wouldn't actually be able to quickly tell when this is true. For example, why do we need to condition on `!$(TargetFramework.Contains('$(_targetOS)'))` and `CustomNativeMain` and why those decide whether we pass `--target` to the linker (try to mentally inline the definition of `_iOSLibraryMode` into places that use the property and whether such definition really makes sense there).

In general - in the places that _use_ the new `_iOSLibraryMode` property, the question is why we're not keying off of `_IsiOSLikePlatform` and `NativeLib` (if necessary - I don't think it's necessary to key off the latter in many of the places that use `_iOSLibraryMode` right now - I think that goes back to Jan's hypothetical "what if we're building an exe" scenario - we don't need to have that scenario but it actually makes the conditions simpler to reason about because it makes it clear what part are due to platform differences and what parts are due to output type differences).

Reply:
Thank you all for you feedback and I understand the concerns.

The specific conditions (like `!$(TargetFramework.Contains('$(_targetOS)'))`) were introduced to avoid looking up the SDK and resolving relevant properties twice (when the build actually pulls in Xamarin SDK). 

I will try to simplify this.

Reply:
How expensive is running the commands? At the end of the day we're going to run an aot compiler so if it's not costly in relative terms to aot compilation I would not try to avoid it too hard.

I now understand that this is trying to approximate the exact conditions under which xamarin sdk or appbuilder would take over but that's also what makes it feel a bit like mystery code.

Reply:
I have simplified the logic as suggested. Please give it another look.